// lnd-grpc/dist/grpc.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _events = _interopRequireDefault(require("events"));
var _javascriptStateMachine = _interopRequireDefault(require("javascript-state-machine"));
var _debug = _interopRequireDefault(require("debug"));
var _parse = _interopRequireDefault(require("lndconnect/parse"));
var _grpcJs = require("@grpc/grpc-js");
var _utils = require("./utils");
var _services = require("./services");
var _registry = _interopRequireDefault(require("./registry"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
const debug = (0, _debug.default)('lnrpc:grpc');

// Set up SSL with the cypher suits that we need.
if (!process.env.GRPC_SSL_CIPHER_SUITES) {
  process.env.GRPC_SSL_CIPHER_SUITES = _utils.grpcSslCipherSuites;
}

/**
 * Lnd gRPC service wrapper.
 * @extends EventEmitter
 */
let LndGrpc = /*#__PURE__*/function (_EventEmitter) {
  _inherits(LndGrpc, _EventEmitter);
  var _super = _createSuper(LndGrpc);
  function LndGrpc(options = {}) {
    var _this;
    _classCallCheck(this, LndGrpc);
    _this = _super.call(this);
    debug(`Initializing LndGrpc with config: %o`, options);
    _this.options = options;

    // If an lndconnect uri was provided, extract the connection details from that.
    if (options.lndconnectUri) {
      const connectionInfo = (0, _parse.default)(options.lndconnectUri);
      Object.assign(_this.options, connectionInfo);
    }

    // Define state machine.
    _this.fsm = new _javascriptStateMachine.default({
      init: 'ready',
      transitions: [{
        name: 'activateWalletUnlocker',
        from: ['ready', 'active'],
        to: 'locked'
      }, {
        name: 'activateLightning',
        from: ['ready', 'locked'],
        to: 'active'
      }, {
        name: 'disconnect',
        from: ['locked', 'active'],
        to: 'ready'
      }],
      methods: {
        onBeforeActivateWalletUnlocker: _this.onBeforeActivateWalletUnlocker.bind(_assertThisInitialized(_this)),
        onBeforeActivateLightning: _this.onBeforeActivateLightning.bind(_assertThisInitialized(_this)),
        onBeforeDisconnect: _this.onBeforeDisconnect.bind(_assertThisInitialized(_this)),
        onAfterDisconnect: _this.onAfterDisconnect.bind(_assertThisInitialized(_this)),
        onInvalidTransition: _utils.onInvalidTransition,
        onPendingTransition: _utils.onPendingTransition
      },
      onInvalidTransition(transition, from, to) {
        throw Object.assign(new Error(`transition is invalid in current state`), {
          transition,
          from,
          to
        });
      }
    });

    // Define services.
    _this.supportedServices = [_services.WalletUnlocker, _services.Lightning, _services.Autopilot, _services.ChainNotifier, _services.Invoices, _services.Router, _services.Signer, _services.State, _services.Versioner, _services.WalletKit, _services.Watchtower, _services.WatchtowerClient];
    _this.services = {};
    _this.tor = (0, _utils.tor)();

    // Instantiate services.
    _this.supportedServices.forEach(Service => {
      const instance = new Service(_this.options);
      _this.services[instance.serviceName] = instance;
    });
    return _this;
  }

  // ------------------------------------
  // FSM Proxies
  // ------------------------------------
  _createClass(LndGrpc, [{
    key: "is",
    value: function is(...args) {
      return this.fsm.is(...args);
    }
  }, {
    key: "can",
    value: function can(...args) {
      return this.fsm.can(...args);
    }
  }, {
    key: "observe",
    value: function observe(...args) {
      return this.fsm.observe(...args);
    }
  }, {
    key: "state",
    get: function () {
      return this.fsm.state;
    }
  }, {
    key: "connect",
    value: async function connect() {
      debug(`Connecting to lnd gRPC service`);

      // Verify that the host is valid.
      const host = this.options.host;
      await (0, _utils.validateHost)(host);

      // Start tor service if needed.
      if ((0, _utils.isTor)(host) && !this.tor.isStarted()) {
        this.emit('tor.starting');
        await this.tor.start();
        this.emit('tor.started');
      }

      // For lnd >= 0.13.*, the state service is available which provides with
      // the wallet state. For lower version of lnd continue to use WalletUnlocker
      // error codes
      let walletState;
      this.isStateServiceAvailable = false;
      try {
        // Subscribe to wallet state and get current state
        let state = await this.getWalletState();
        if (state == 'WAITING_TO_START') {
          state = await this.checkWalletState(['NON_EXISTING', 'LOCKED']);
        }
        switch (state) {
          case 'NON_EXISTING':
          case 'LOCKED':
            walletState = _utils.WALLET_STATE_LOCKED;
            break;
          case 'UNLOCKED':
            // Do nothing.
            break;
          case 'RPC_ACTIVE':
          case 'SERVER_ACTIVE':
            walletState = _utils.WALLET_STATE_ACTIVE;
            break;
        }
        this.isStateServiceAvailable = true;
      } catch (error) {
        if (error.code === _grpcJs.status.UNIMPLEMENTED) {
          // Probe the services to determine the wallet state.
          walletState = await this.determineWalletState();
        }
      }
      switch (walletState) {
        case _utils.WALLET_STATE_LOCKED:
          await this.activateWalletUnlocker();
          break;
        case _utils.WALLET_STATE_ACTIVE:
          await this.activateLightning();
          break;
      }
    }
  }, {
    key: "activateWalletUnlocker",
    value: async function activateWalletUnlocker(...args) {
      await this.fsm.activateWalletUnlocker(...args);
      this.emit('locked');
    }
  }, {
    key: "activateLightning",
    value: async function activateLightning(...args) {
      try {
        await this.fsm.activateLightning(...args);
        this.emit('active');
      } catch (e) {
        await this.disconnectAll();
        throw e;
      }
    }
  }, {
    key: "disconnect",
    value: async function disconnect(...args) {
      if (this.can('disconnect')) {
        await this.fsm.disconnect(...args);
      }
      if (this.tor.isStarted()) {
        this.emit('tor.stopping');
        await this.tor.stop();
        this.emit('tor.stopped');
      }
      this.emit('disconnected');
    }

    // ------------------------------------
    // FSM Observers
    // ------------------------------------

    /**
     * Disconnect from the gRPC service.
     */
  }, {
    key: "onBeforeDisconnect",
    value: async function onBeforeDisconnect() {
      debug(`Disconnecting from lnd gRPC service`);
      await this.disconnectAll();
    }
    /**
     * Log successful disconnect.
     */
  }, {
    key: "onAfterDisconnect",
    value: async function onAfterDisconnect() {
      debug('Disconnected from lnd gRPC service');
    }

    /**
     * Connect to and activate the wallet unlocker api.
     */
  }, {
    key: "onBeforeActivateWalletUnlocker",
    value: async function onBeforeActivateWalletUnlocker() {
      if (this.services.WalletUnlocker.can('connect')) {
        await this.services.WalletUnlocker.connect();
      }
    }

    /**
     * Connect to and activate the main api.
     */
  }, {
    key: "onBeforeActivateLightning",
    value: async function onBeforeActivateLightning() {
      const _this$services = this.services,
        Lightning = _this$services.Lightning,
        WalletUnlocker = _this$services.WalletUnlocker;

      // await for RPC_ACTIVE or SERVER_ACTIVE state before interacting if needed
      if (this.isStateServiceAvailable) {
        await this.checkWalletState(['RPC_ACTIVE', 'SERVER_ACTIVE']);
      }

      // Disconnect wallet unlocker if its connected.
      if (WalletUnlocker.can('disconnect')) {
        await WalletUnlocker.disconnect();
      }
      // First connect to the Lightning service.
      await Lightning.connect();

      // Fetch the determined version.
      const version = Lightning.version;

      // Get a list of all other available and supported services.
      const availableServices = _registry.default[version].services.map(s => s.name).filter(s => Object.keys(this.services).includes(s)).filter(s => !['WalletUnlocker', 'Lightning'].includes(s));

      // Connect to the other services.
      await Promise.all(availableServices.filter(serviceName => this.services[serviceName].can('connect')).map(serviceName => {
        const service = this.services[serviceName];
        service.version = version;
        // Disable waiting for cert/macaroon for sub-services.
        return service.connect({
          waitForCert: false,
          waitForMacaroon: false
        });
      }));
    }

    // ------------------------------------
    // Helpers
    // ------------------------------------

    /**
     * Disconnect all services.
     */
  }, {
    key: "disconnectAll",
    value: async function disconnectAll() {
      debug('Disconnecting from all gRPC services');
      await Promise.all(Object.keys(this.services).map(serviceName => {
        const service = this.services[serviceName];
        if (service.can('disconnect')) {
          return service.disconnect();
        }
      }));
      debug('Disconnected from all gRPC services');
    }

    /**
     * Probe to determine what state lnd is in.
     */
  }, {
    key: "determineWalletState",
    value: async function determineWalletState(options = {
      keepalive: false
    }) {
      debug('Attempting to determine wallet state');
      let walletState;
      try {
        await this.services.WalletUnlocker.connect();
        // Call the unlockWallet method with a missing password argument.
        // This is a way of probing the api to determine it's state.
        await this.services.WalletUnlocker.unlockWallet();
      } catch (error) {
        switch (error.code) {
          /*
            `UNIMPLEMENTED` indicates that the requested operation is not implemented or not supported/enabled in the
             service. This implies that the wallet is already unlocked, since the WalletUnlocker service is not active.
             See
              `DEADLINE_EXCEEDED` indicates that the deadline expired before the operation could complete. In the case of
             our probe here the likely cause of this is that we are connecting to an lnd node where the `noseedbackup`
             flag has been set and therefore the `WalletUnlocker` interace is non-functional.
              https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/constants.js#L129.
           */
          case _grpcJs.status.UNIMPLEMENTED:
          case _grpcJs.status.DEADLINE_EXCEEDED:
            debug('Determined wallet state as:', _utils.WALLET_STATE_ACTIVE);
            walletState = _utils.WALLET_STATE_ACTIVE;
            return walletState;

          /**
            `UNKNOWN` indicates that unlockWallet was called without an argument which is invalid.
            This implies that the wallet is waiting to be unlocked.
          */
          case _grpcJs.status.UNKNOWN:
            debug('Determined wallet state as:', _utils.WALLET_STATE_LOCKED);
            walletState = _utils.WALLET_STATE_LOCKED;
            return walletState;

          /**
            Bubble all other errors back to the caller and abort the connection attempt.
            Disconnect all services.
          */
          default:
            debug('Unable to determine wallet state', error);
            throw error;
        }
      } finally {
        if (!options.keepalive && this.can('disconnect')) {
          await this.disconnect();
        }
      }
    }

    /**
     * Wait for wallet to enter a particular state.
     * @param  {string} state state to wait for (RPC_ACTIVE, LOCKED, UNLOCKED)
     * @return {Promise<Object>}.
     */
  }, {
    key: "checkWalletState",
    value: function checkWalletState(states) {
      states = [].concat(states);
      const waitForState = resolve => {
        return this.services.State.getState().then(currentState => {
          debug('Got wallet state as %o', currentState);
          if (states.includes(currentState.state)) {
            resolve(currentState.state);
          } else {
            setTimeout(_ => waitForState(resolve), 400);
          }
        });
      };
      return (0, _utils.promiseTimeout)(_utils.CONNECT_WAIT_TIMEOUT * 1000, new Promise(waitForState), 'Connection timeout out.');
    }

    /**
     * Wait for lnd to enter a particular state.
     * @param  {string} state Name of state to wait for (locked, active, disconnected)
     * @return {Promise<Object>} Object with `isDone` and `cancel` properties.
     */
  }, {
    key: "waitForState",
    value: function waitForState(stateName) {
      let successHandler;

      /**
       * Promise that resolves when service is active.
       */
      const isDone = new Promise(resolve => {
        // If the service is already in the requested state, return immediately.
        if (this.fsm.state === stateName) {
          return resolve();
        }
        // Otherwise, wait until we receive a relevant state change event.
        successHandler = () => resolve();
        this.prependOnceListener(stateName, successHandler);
      });

      /**
       * Method to abort the wait (prevent the isDone from resolving and remove activation event listener).
       */
      const cancel = () => {
        if (successHandler) {
          this.off(stateName, successHandler);
          successHandler = null;
        }
      };
      return {
        isDone,
        cancel
      };
    }

    /**
     * Get current wallet state
     * @return {Promise<string>}.
     */
  }, {
    key: "getWalletState",
    value: async function getWalletState() {
      if (this.services.State.can('connect')) {
        await this.services.State.connect();
      }
      const _await$this$services$ = await this.services.State.getState(),
        state = _await$this$services$.state;
      debug('Got wallet state as %o', state);
      return state;
    }
  }]);
  return LndGrpc;
}(_events.default);
var _default = LndGrpc;
exports.default = _default;
module.exports = exports.default;
