// lnd-grpc/dist/utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WALLET_STATE_LOCKED = exports.WALLET_STATE_ACTIVE = exports.TOR_WAIT_TIMEOUT = exports.SERVICE_CONNECT_TIMEOUT = exports.PROBE_TIMEOUT = exports.PROBE_RETRY_INTERVAL = exports.FILE_WAIT_TIMEOUT = exports.CONNECT_WAIT_TIMEOUT_TOR = exports.CONNECT_WAIT_TIMEOUT = void 0;
// Wallet states.
const WALLET_STATE_LOCKED = 'WALLET_STATE_LOCKED';
exports.WALLET_STATE_LOCKED = WALLET_STATE_LOCKED;
const WALLET_STATE_ACTIVE = 'WALLET_STATE_ACTIVE';

// Time (in seconds) to wait for interface probe calls to complete.
exports.WALLET_STATE_ACTIVE = WALLET_STATE_ACTIVE;
const PROBE_TIMEOUT = 25;

// Time (in ms) to wait before retrying a connection attempt.
exports.PROBE_TIMEOUT = PROBE_TIMEOUT;
const PROBE_RETRY_INTERVAL = 250;

// Time (in seconds) to wait for a grpc connection to be established.
exports.PROBE_RETRY_INTERVAL = PROBE_RETRY_INTERVAL;
const CONNECT_WAIT_TIMEOUT = 15;

// Time (in seconds) to wait for a grpc connection to be established.
exports.CONNECT_WAIT_TIMEOUT = CONNECT_WAIT_TIMEOUT;
const CONNECT_WAIT_TIMEOUT_TOR = 30;

// Time (in seconds) to wait before aborting service connection attempts.
exports.CONNECT_WAIT_TIMEOUT_TOR = CONNECT_WAIT_TIMEOUT_TOR;
const SERVICE_CONNECT_TIMEOUT = 35;

// Time (in ms) to wait for a cert/macaroon file to become present.
exports.SERVICE_CONNECT_TIMEOUT = SERVICE_CONNECT_TIMEOUT;
const FILE_WAIT_TIMEOUT = 10 * 1000;

// Time (in ms) to wait for Tor to become ready after starting.
exports.FILE_WAIT_TIMEOUT = FILE_WAIT_TIMEOUT;
const TOR_WAIT_TIMEOUT = 1000;
exports.TOR_WAIT_TIMEOUT = TOR_WAIT_TIMEOUT;
