// lnd-grpc/dist/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  createMacaroonCreds: true,
  getMacaroon: true,
  createSslCreds: true,
  delay: true,
  getDeadline: true,
  grpcOptions: true,
  grpcSslCipherSuites: true,
  promisifiedCall: true,
  validateHost: true,
  waitForFile: true,
  isTor: true,
  promiseTimeout: true,
  tor: true,
  onInvalidTransition: true,
  onPendingTransition: true
};
Object.defineProperty(exports, "createMacaroonCreds", {
  enumerable: true,
  get: function () {
    return _createMacaroonCreds.default;
  }
});
Object.defineProperty(exports, "createSslCreds", {
  enumerable: true,
  get: function () {
    return _createSslCreds.default;
  }
});
Object.defineProperty(exports, "delay", {
  enumerable: true,
  get: function () {
    return _delay.default;
  }
});
Object.defineProperty(exports, "getDeadline", {
  enumerable: true,
  get: function () {
    return _getDeadline.default;
  }
});
Object.defineProperty(exports, "getMacaroon", {
  enumerable: true,
  get: function () {
    return _createMacaroonCreds.getMacaroon;
  }
});
Object.defineProperty(exports, "grpcOptions", {
  enumerable: true,
  get: function () {
    return _grpcOptions.default;
  }
});
Object.defineProperty(exports, "grpcSslCipherSuites", {
  enumerable: true,
  get: function () {
    return _grpcSslCipherSuites.default;
  }
});
Object.defineProperty(exports, "isTor", {
  enumerable: true,
  get: function () {
    return _isTor.default;
  }
});
Object.defineProperty(exports, "onInvalidTransition", {
  enumerable: true,
  get: function () {
    return _stateMachineErrorHandlers.onInvalidTransition;
  }
});
Object.defineProperty(exports, "onPendingTransition", {
  enumerable: true,
  get: function () {
    return _stateMachineErrorHandlers.onPendingTransition;
  }
});
Object.defineProperty(exports, "promiseTimeout", {
  enumerable: true,
  get: function () {
    return _promiseTimeout.default;
  }
});
Object.defineProperty(exports, "promisifiedCall", {
  enumerable: true,
  get: function () {
    return _promisifiedCall.default;
  }
});
Object.defineProperty(exports, "tor", {
  enumerable: true,
  get: function () {
    return _tor.default;
  }
});
Object.defineProperty(exports, "validateHost", {
  enumerable: true,
  get: function () {
    return _validateHost.default;
  }
});
Object.defineProperty(exports, "waitForFile", {
  enumerable: true,
  get: function () {
    return _waitForFile.default;
  }
});
var _createMacaroonCreds = _interopRequireWildcard(require("./createMacaroonCreds"));
var _createSslCreds = _interopRequireDefault(require("./createSslCreds"));
var _delay = _interopRequireDefault(require("./delay"));
var _getDeadline = _interopRequireDefault(require("./getDeadline"));
var _grpcOptions = _interopRequireDefault(require("./grpcOptions"));
var _grpcSslCipherSuites = _interopRequireDefault(require("./grpcSslCipherSuites"));
var _promisifiedCall = _interopRequireDefault(require("./promisifiedCall"));
var _validateHost = _interopRequireDefault(require("./validateHost"));
var _waitForFile = _interopRequireDefault(require("./waitForFile"));
var _isTor = _interopRequireDefault(require("./isTor"));
var _promiseTimeout = _interopRequireDefault(require("./promiseTimeout"));
var _tor = _interopRequireDefault(require("./tor"));
var _stateMachineErrorHandlers = require("./stateMachineErrorHandlers");
var _constants = require("./constants");
Object.keys(_constants).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _constants[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _constants[key];
    }
  });
});
var _proto = require("./proto");
Object.keys(_proto).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _proto[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _proto[key];
    }
  });
});
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
