// lnd-grpc/dist/utils/proto.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getProtoVersions = exports.getProtoDir = exports.getLatestProtoVersion = exports.getClosestProtoVersion = exports.default = exports.GRPC_LOWEST_VERSION = exports.GRPC_HIGHEST_STABLE_VERSION = void 0;
var _path = require("path");
var _semver = _interopRequireDefault(require("semver"));
var _debug = _interopRequireDefault(require("debug"));
var _registry = _interopRequireDefault(require("../registry"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
const debug = (0, _debug.default)('lnrpc:proto');
const GRPC_LOWEST_VERSION = '0.4.2-beta';
exports.GRPC_LOWEST_VERSION = GRPC_LOWEST_VERSION;
const GRPC_HIGHEST_STABLE_VERSION = '0.14.4-beta';

/**
 * Get the directory where rpc.proto files are stored.
 * @return {String} Directory where rpoc.proto files are stored.
 */
exports.GRPC_HIGHEST_STABLE_VERSION = GRPC_HIGHEST_STABLE_VERSION;
const getProtoDir = () => {
  return (0, _path.join)(__dirname, '..', '..', 'proto');
};

/**
 * Get a list of all rpc.proto versions that we provide.
 * @return {Promise<Array>} List of available rpc.proto versions.
 */
exports.getProtoDir = getProtoDir;
const getProtoVersions = () => {
  return Object.keys(_registry.default);
};

/**
 * Get the latest rpc.proto version that we provide.
 * @return {Promise<String>} The latest rpc.proto version that we provide.
 */
exports.getProtoVersions = getProtoVersions;
const getLatestProtoVersion = () => {
  const versions = getProtoVersions();
  return _semver.default.maxSatisfying(versions, `> ${GRPC_LOWEST_VERSION}`, {
    includePrerelease: true
  });
};

/**
 * Find the closest supported rpc.proto version based on an lnd version string.
 * @param  {[type]}  info [description]
 * @return {Promise}      [description]
 */
exports.getLatestProtoVersion = getLatestProtoVersion;
const getClosestProtoVersion = async versionString => {
  debug('Testing version string: %s', versionString);
  let _versionString$split = versionString.split(' '),
    _versionString$split2 = _slicedToArray(_versionString$split, 2),
    version = _versionString$split2[0],
    commitString = _versionString$split2[1];
  debug('Parsed version string into version: %s, commitString: %s', version, commitString);

  // If this looks like a pre-release.
  if (version.endsWith('99-beta')) {
    throw new Error(`Unsupported prerelease version: ${versionString}`);
  }
  const supportedVersions = getProtoVersions();
  debug('Searching for closest match for version %s in range: %o', version, supportedVersions);
  let match = _semver.default.maxSatisfying(supportedVersions, `<= ${version}`, {
    includePrerelease: true
  });
  debug('Determined closest rpc.proto match as: %s', match);
  return match;
};
exports.getClosestProtoVersion = getClosestProtoVersion;
var _default = {
  getProtoDir,
  getProtoVersions,
  getLatestProtoVersion,
  getClosestProtoVersion
};
exports.default = _default;
